### PR TITLE
chore: create the basis for unit testing framework for covering the project with tests

### PR DESCRIPTION
**Pull request by bozerkins**
### Description

This is the start of an argo-watcher server refactoring, with an end goal of **having better error handling code**. 

Before going straight to the error handling changes, I realized that last time I tested them on a real production `ArgoCD` instance, running real applications and failing some deployments manually. I would like to avoid using this approach, and instead would prefer having **local unit tests that can be run on demand**. 

### Approach

To have local unit tests for communication with `ArgoCD`, we need to make several improvements. 

1) Different module configurations (argo, state, router) should not rely on ENV variables, but rather have a centralized config object that can be changed for unit tests
2) A way to inject mock objects to our modules, so we can test different scenarios in the isolated way (locally)
3) Move API calls to `ArgoCD` into a separate module, so it can be mocked and tested without `cmd/mock` application

Reference to an approach that I think will work for us: https://www.alexedwards.net/blog/organising-database-access

---
### What was done

**Scope**: Argo Watcher Server

I tried to make minimal changes to the client, since at this moment it's outside of scope of the refactoring. The goal of the refactoring is improving testability of the code.

I ended up refactoring the server in the following ways
1. I removed reliance on `environment variables` and `global variables` from most of the code.
2. Divide functionality into smaller files where it made sense. 
2.1. Methods that were called as controller functions from router i moved to `router. go` file
2.2. Created a separate package for server configurations, which reads the environment variables and can be mocked `config/config.go`
2.3. Moved calls to `ArgoCD API` into a separate `argo_api.go` file, so that API calls could be mocked
2.4. Moved functionality that tracks `ArgoCD` application status in a go routine into a separate module `argo_status_updater.go`. Moved the creation of this go routine from Argo class to the router function. This needed to be done to ensure testability of that code. We can now write tests to make sure the go routine works as expected.
2.5. While writing tests for `Argo` class, I realized that we don't handle database errors. I added some errors handling for database state and changed the interface slightly, but that's only the beginning. We need to handle database errors properly. At this moment most of the database errors cause Argo Watcher to panic. This [programming principle](https://97-things-every-x-should-know.gitbook.io/97-things-every-programmer-should-know/en/thing_26) nicely covers the approach
3. I created unit tests for functions `Argo.Check` and `Argo.AddTask` in `argo_test.go` file. Those tests are proper unit tests, that mock API calls, Metric collection, and State management (database). The tests do not depend on ArgoCD, Database or any Prometheus process. The tests describe the expected behavior and can be used to ensure the changes to the code base do not break anything.
4. I removed short aliases for the modules (c for config, m for models, h for helpers), since it's considered a better practice to use full names in the code. This [programming principle](https://97-things-every-x-should-know.gitbook.io/97-things-every-programmer-should-know/en/thing_15) nicely covers the reasoning for descriptive variables.

### Further work

Using `argo_test.go` as an example, we need to create tests for the rest of the argo-watcher codebase. @shini4i, it'd be nice for us to do that in smaller PRs (this one is way to big, imho). While doing the tests, we will realize there are a lot of corner cases that we don't cover, for example database errors.

The rule of thumb for writing the tests is: if you can't mock the execution and thoroughly test it, that means the code is too intertwined and needs to be divided further into classes / modules / functions.

After we're doing with writing tests, I think we'd need to add test coverage reports and enforce test coverage for any further changes.

After we're done with unit tests, we can create integration tests too. We could set up a small test environment that could test our API with Argo Mock (or real Argo) and real database.

> As always, comments are welcome :)

### CODE REVIEW

File: cmd/argo-watcher/argo_api.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}(resp.Body)
+
+	if resp.StatusCode != 200 {
+		var argoErrorResponse models.ArgoApiErrorResponse
+		if err = json.Unmarshal(body, &argoErrorResponse); err != nil {
+			return nil, fmt.Errorf("could not parse json error response: %s", body)
+		}
+
+		if argoErrorResponse.Message == "" {
```
Comment (shini4i) [inline cmd/argo-watcher/argo_api.go:143]: It seems that this "if" block is redundant

File: cmd/argo-watcher/argo_api.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+				"failed parsing argocd API response: %s", 
+				bytes.NewBuffer(body).String(),
+			)
+		}
+
+		return nil, errors.New(argoErrorResponse.Message)
+	}
+
+	var argoApp models.Application
+	err = json.Unmarshal(body, &argoApp)
```
Comment (shini4i) [inline cmd/argo-watcher/argo_api.go]: Here we have a different approach for handling errors in "json.Umarshal". Let's make it the same as on line 139.

File: cmd/argo-watcher/argo_status_updater.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	if status == ArgoAppFailed {
+		updater.handleArgoAPIFailure(task, err)
+		return
+	}
+
+	// fetch application details
+	app, err := updater.argo.api.GetApplication(task.App)
+
+	// define default message
+	const defaultErrorMessage string = "could not retrieve details"
```
Comment (shini4i) [inline cmd/argo-watcher/argo_status_updater.go]: Aren't we redefining this constant every time? Let's move it outside of the function body.

File: cmd/argo-watcher/argo_status_updater.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	switch status {
+	// not all images were deployed to the application
+	case ArgoAppNotAvailable:
+		// show list of missing images
+		var message string
+		// define details
+		if err != nil {
+			message = defaultErrorMessage
+		} else {
+			message = "List of current images (last app check):\n"
```
Comment (shini4i) [inline cmd/argo-watcher/argo_status_updater.go]: Maybe let's replace multiple "message +=" with a single fmt.Sprintf?

File: cmd/argo-watcher/argo_status_updater.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		updater.handleAppNotAvailable(task, errors.New(message))
+	// application sync status wasn't valid
+	case ArgoAppNotSynced:
+		// display sync status and last sync message
+		var message string
+		// define details
+		if err != nil {
+			message = defaultErrorMessage
+		} else {
+			message = "App status \"" + app.Status.OperationState.Phase + "\"\n"
```
Comment (shini4i) [inline cmd/argo-watcher/argo_status_updater.go]: same here

File: cmd/argo-watcher/argo_status_updater.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+		updater.handleAppOutOfSync(task, errors.New(message))
+	// application is not in a healthy status
+	case ArgoAppNotHealthy:
+		// display current health of pods
+		var message string
+		// define details
+		if err != nil {
+			message = defaultErrorMessage
+		} else {
+			message = "App sync status \"" + app.Status.Sync.Status + "\"\n"
```
Comment (shini4i) [inline cmd/argo-watcher/argo_status_updater.go]: and here

File: cmd/argo-watcher/argo_api.go (hunk 1 above)
Comment (bozerkins) [reply cmd/argo-watcher/argo_api.go:143]: I think it's not redundant. It's possible that we successfully parse the response, but the message we return on the next line is empty. In that case, without the check present, we would return an exception with an empty message and thus provide no details about the error. 

The check ensures that if there's no error message, we return the whole `json string` and the user can decide what the response is about. 

In the future - I think it's definitely need to do a small investigation into what argo API returns and clean this up, because the whole assumption is that it's possible to have empty message (which I'm not sure is true, but I think it is). This IF statement is copy-pasted.

File: cmd/argo-watcher/argo_api.go (hunk 2 above)
Comment (bozerkins) [reply cmd/argo-watcher/argo_api.go]: Good point. Will change that

File: cmd/argo-watcher/argo_status_updater.go (hunk 4 above)
Comment (bozerkins) [reply cmd/argo-watcher/argo_status_updater.go]: I'll try to make it look better

**Review by shini4i — APPROVED** [review]
